### PR TITLE
RBAC-1

### PR DESCRIPTION
**Pull request by vitlibar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Improvement

Just refactoring in order to implement RBAC, no new functionality here.

### CODE REVIEW

File: dbms/src/Access/AllowedClientHosts.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        String toString() const;
+
+        friend bool operator ==(const IPSubnet & lhs, const IPSubnet & rhs) { return (lhs.prefix == rhs.prefix) && (lhs.mask == rhs.mask); }
+        friend bool operator !=(const IPSubnet & lhs, const IPSubnet & rhs) { return !(lhs == rhs); }
+    };
+
+    struct AllAddressesTag {};
+
+    AllowedClientHosts();
+    AllowedClientHosts(AllAddressesTag);
```
Comment (abyss7) [inline dbms/src/Access/AllowedClientHosts.h]: `explicit`

File: dbms/src/Access/AllowedClientHosts.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    /// Checks if the provided address is in the list. Returns false if not.
+    bool contains(const IPAddress & address) const;
+
+    /// Checks if any address is allowed.
+    bool containsAllAddresses() const;
+
+    /// Checks if the provided address is in the list. Throws an exception if not.
+    /// `username` is only used for generating an error message if the address isn't in the list.
+    void checkContains(const IPAddress & address, const String & user_name = String()) const;
```
Comment (abyss7) [inline dbms/src/Access/AllowedClientHosts.h]: I suggest to remove this method from interface - passing `user_name` can be replaced with outer call to `contains()` where the username is known.

Comment (vitlibar) [reply dbms/src/Access/AllowedClientHosts.h]: > I suggest to remove this method from interface - passing `user_name` can be replaced with outer call to `contains()` where the username is known.

The code generating an error message for that purpose isn't very simple, and I'd like to keep this kind of details inside this specific `AllowedClientHosts` class, not inside the more generic class `User` (which has plenty other work to do).

File: dbms/src/Access/Authentication.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    void setPassword(const String & password);
+
+    /// Returns the password. Allowed to use only for Type::PLAINTEXT_PASSWORD.
+    String getPassword() const;
+
+    /// Sets the password as a string of hexadecimal digits.
+    void setPasswordHashHex(const String & hash);
+    String getPasswordHashHex() const;
+
+    /// Sets the password in binary form.
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply dbms/src/Access/Authentication.h]: > Looks like unneeded for public interface.
The public function `getPasswordHashBinary() is already used by `MySQLProtocol`. I suppose the `setPasswordHashBinary()` can be public too for consistency.

File: dbms/src/Access/AllowedClientHosts.h (hunk 1 above)
Comment (vitlibar) [reply dbms/src/Access/AllowedClientHosts.h]: done

File: dbms/src/Access/Authentication.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    using Digest = std::vector<UInt8>;
+
+    Authentication(Authentication::Type type = NO_PASSWORD);
+    Authentication(const Authentication & src) = default;
+    Authentication & operator =(const Authentication & src) = default;
+    Authentication(Authentication && src) = default;
+    Authentication & operator =(Authentication && src) = default;
+
+    void setType(Type type_) { type = type_; }
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply dbms/src/Access/Authentication.h]: done

**Pull request merged**